scripts/country-uk/ordnance.py

```diff
@@ -0,0 +1,288 @@
+import shapefile
+import csv
+import re
+
+# Note: This requires the UK Ordnance Survey Electoral Boundaries dataset,
+# which is free and liberally licensed, but requires agreeing to a license and
+# downloading via an emailed link from their site.
+# https://www.ordnancesurvey.co.uk/opendatadownload/products.html#BDLINE
+
+# it also assumes pyshp -- pip3 install pyshp
+
+# Change this to your download location
+data_dir = 'bdline_essh_gb/Data'
+uk_dir = '{}/GB'.format(data_dir)
+wales_dir = '{}/Wales'.format(data_dir)
+cerem_dir = '{}/Supplementary_Ceremonial'.format(data_dir)
+
+# the field list for each data set can be quickly viewed using the get_overview function,
+# and debug_print dumps the data to console
+
+
+def read_records(filename):
+    sf = shapefile.Reader(filename)
+    records = sf.records()
+    for record in records:
+        yield record
+
+
+def debug_print(filename):
+    for record in read_records(filename):
+        print(record)
+
+
+def get_overview(filename):
+    sf = shapefile.Reader(filename)
+    return sf.fields
+
+
+def write_csv(filename, csv_columns, dict_data):
+    with open(filename, 'w') as csv_file:
+        writer = csv.DictWriter(csv_file, fieldnames=csv_columns)
+        writer.writeheader()
+        for data in dict_data:
+            writer.writerow(data)
+
+
+def make_id(type_id):
+    replacements = ['\'', ' &', ',', ' Assembly Const',
+                    ' P Const', ' GL', ' ED', ' CP', '_CONST', '(', ')', ]
+    for replacement in replacements:
+        type_id = type_id.replace(replacement, '')
+
+    regex_replacements = [r' County$',
+                          r'_county$',
+                          r' District$',
+                          r'_district$',
+                          r' Authority$',
+                          r'_authority$',
+                          ]
+    for replacement in regex_replacements:
+        type_id = re.sub(replacement, '', type_id, flags=re.IGNORECASE)
+
+    type_id = type_id.lower()
+    type_id = re.sub('\.? ', '_', type_id)
+    type_id = re.sub('[^\w0-9~_.-]', '~', type_id, re.UNICODE)
```

Comment:
I'd like to understand more about the slugging algorithm used here. To what extent is this mandated by the spec/project convention/individual preference?

Reply:
[Pulled from the spec](http://docs.opencivicdata.org/en/latest/proposals/0002.html#identifier-scheme). I believe I took the regex from [elsewhere in the project](https://github.com/opencivicdata/ocd-division-ids/blob/master/scripts/country-us/census_places.py#L167) but it's been a few months so i'm not sure.

The replacements were an attempt not add redundant data (to avoid things like ocd-division/country:uk/county:sample/sample_county) and to smooth out some stylistic choices that seemed to be Ordnance specific.

If you catch any cases where i'm grossly mishandling special characters or other abbrevs, or this ID scheme is doing something that makes no sense to a local user, please let me know.